Cleanup duplicate barcodes

seichhorn opened the pull request:
Adding parameter to barcode filtering methods that removes barcodes that are putative duplicates in adjacent z planes, leaving just one instance of any duplicated barcodes.

pep8speaks commented:
Hello @seichhorn! Thanks for opening this PR.

* In the file [`merlin/analysis/filterbarcodes.py`](https://github.com/emanuega/MERlin/blob/703282cf98baa4f69197b0a10f86c39be770ae91/merlin/analysis/filterbarcodes.py):

> [Line 55:81](https://github.com/emanuega/MERlin/blob/703282cf98baa4f69197b0a10f86c39be770ae91/merlin/analysis/filterbarcodes.py#L55): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 80 characters)
> [Line 74:81](https://github.com/emanuega/MERlin/blob/703282cf98baa4f69197b0a10f86c39be770ae91/merlin/analysis/filterbarcodes.py#L74): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 80 characters)
> [Line 387:81](https://github.com/emanuega/MERlin/blob/703282cf98baa4f69197b0a10f86c39be770ae91/merlin/analysis/filterbarcodes.py#L387): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (95 > 80 characters)

* In the file [`merlin/util/duplicatebcremoval.py`](https://github.com/emanuega/MERlin/blob/703282cf98baa4f69197b0a10f86c39be770ae91/merlin/util/duplicatebcremoval.py):

> [Line 7:1](https://github.com/emanuega/MERlin/blob/703282cf98baa4f69197b0a10f86c39be770ae91/merlin/util/duplicatebcremoval.py#L7): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 7:81](https://github.com/emanuega/MERlin/blob/703282cf98baa4f69197b0a10f86c39be770ae91/merlin/util/duplicatebcremoval.py#L7): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 80 characters)
> [Line 8:81](https://github.com/emanuega/MERlin/blob/703282cf98baa4f69197b0a10f86c39be770ae91/merlin/util/duplicatebcremoval.py#L8): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (111 > 80 characters)
> [Line 9:81](https://github.com/emanuega/MERlin/blob/703282cf98baa4f69197b0a10f86c39be770ae91/merlin/util/duplicatebcremoval.py#L9): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (110 > 80 characters)
> [Line 10:81](https://github.com/emanuega/MERlin/blob/703282cf98baa4f69197b0a10f86c39be770ae91/merlin/util/duplicatebcremoval.py#L10): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (109 > 80 characters)
> [Line 11:81](https://github.com/emanuega/MERlin/blob/703282cf98baa4f69197b0a10f86c39be770ae91/merlin/util/duplicatebcremoval.py#L11): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (111 > 80 characters)
> [Line 12:81](https://github.com/emanuega/MERlin/blob/703282cf98baa4f69197b0a10f86c39be770ae91/merlin/util/duplicatebcremoval.py#L12): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (110 > 80 characters)
> [Line 16:81](https://github.com/emanuega/MERlin/blob/703282cf98baa4f69197b0a10f86c39be770ae91/merlin/util/duplicatebcremoval.py#L16): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 80 characters)
> [Line 17:81](https://github.com/emanuega/MERlin/blob/703282cf98baa4f69197b0a10f86c39be770ae91/merlin/util/duplicatebcremoval.py#L17): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 80 characters)
> [Line 18:81](https://github.com/emanuega/MERlin/blob/703282cf98baa4f69197b0a10f86c39be770ae91/merlin/util/duplicatebcremoval.py#L18): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (114 > 80 characters)
> [Line 20:81](https://github.com/emanuega/MERlin/blob/703282cf98baa4f69197b0a10f86c39be770ae91/merlin/util/duplicatebcremoval.py#L20): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (113 > 80 characters)
> [Line 23:30](https://github.com/emanuega/MERlin/blob/703282cf98baa4f69197b0a10f86c39be770ae91/merlin/util/duplicatebcremoval.py#L23): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 23:32](https://github.com/emanuega/MERlin/blob/703282cf98baa4f69197b0a10f86c39be770ae91/merlin/util/duplicatebcremoval.py#L23): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 23:46](https://github.com/emanuega/MERlin/blob/703282cf98baa4f69197b0a10f86c39be770ae91/merlin/util/duplicatebcremoval.py#L23): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 23:48](https://github.com/emanuega/MERlin/blob/703282cf98baa4f69197b0a10f86c39be770ae91/merlin/util/duplicatebcremoval.py#L23): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 35:81](https://github.com/emanuega/MERlin/blob/703282cf98baa4f69197b0a10f86c39be770ae91/merlin/util/duplicatebcremoval.py#L35): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 80 characters)
> [Line 39:81](https://github.com/emanuega/MERlin/blob/703282cf98baa4f69197b0a10f86c39be770ae91/merlin/util/duplicatebcremoval.py#L39): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 80 characters)
> [Line 40:81](https://github.com/emanuega/MERlin/blob/703282cf98baa4f69197b0a10f86c39be770ae91/merlin/util/duplicatebcremoval.py#L40): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (109 > 80 characters)
> [Line 42:1](https://github.com/emanuega/MERlin/blob/703282cf98baa4f69197b0a10f86c39be770ae91/merlin/util/duplicatebcremoval.py#L42): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

emanuega commented on docs/tasks.rst:
```diff
@@ -87,6 +91,9 @@ Description: Use an adaptive barcode to enrich the decode barcodes for the corre
+* remove\_zplane\_duplicates -- Remove putative duplicate barcode counts from adjacent z planes.
+* z\_planes\_above\_below -- If removing putative duplicate barcodes, number of adjacent z planes to consider, generally anything within 2 µm would be worth considering.
+* max\_centroid\_separation -- If removing putative duplicate barcodes, maximum euclidean distance in xy pixels that can separate the centroids of putative duplicates.
 
```
Can we rename these parameters so it is clearer that they relate to the removal of duplicates across the z planes. Maybe "remove_z_duplicated_barcodes" "z_duplicate_axial_threshold" and "z_duplicate_lateral_threshold"

emanuega replied:
Looks like these parameters were renamed in the code and just have to be updated in the documentation.

seichhorn replied:
Yeah forgot them there. Should be good now.

emanuega commented on merlin/util/duplicatebcremoval.py:
```diff
@@ -0,0 +1,53 @@
+import numpy as np
+from scipy.spatial import cKDTree
+import networkx as nx
+import pandas as pd
```
Can we rename this module so that in the future we could move some of the adaptive thresholding code into this file?

emanuega commented on merlin/util/duplicatebcremoval.py:
```diff
@@ -0,0 +1,53 @@
+        -> pd.DataFrame:
+    """ Depending on the separation between z planes, spots from a single
+        molecule may be observed in more than one z plane. These putative
+        duplicates are removed based on supplied distance and z plane
```
It could be useful to have a unit test for this function to make sure it behaves correctly in the edge cases and that we don't break it in the future.

emanuega commented on merlin/util/duplicatebcremoval.py:
```diff
@@ -0,0 +1,53 @@
+                                   distance_upper_bound=maxDist)
+            currentHits = treeBC.index.values[idx[np.isfinite(dist)]]
+            comparisonHits = queryBC.index.values[np.isfinite(dist)]
+            graph.add_edges_from(list(zip(currentHits, comparisonHits)))
```
I would expect this to only remove the barcodes assigned to the same barcode_id, but I don't find anywhere where barcode_id is being checked. Is this the correct assumption?

seichhorn replied:
In the implementation in filter barcodes the barcode database is loaded and then grouped by barcode id, the groups are run through util.duplicatebcremoval.cleanup_across_z individually so it would only remove barcodes assigned to the same id. I didn't put any logic explicitly enforcing this within the cleanup_across_z function, but I could put a check or warning or something if the user decided to run it on their own and didn't preprocess the data correctly?

emanuega replied:
Sorry, I meant to delete this comment when I added the renaming suggesting back in.

emanuega commented on merlin/util/duplicatebcremoval.py:
```diff
@@ -0,0 +1,53 @@
+            graph.add_edges_from(list(zip(currentHits, comparisonHits)))
+    connectedComponents = [list(x) for x in
+                           list(nx.connected_components(graph))]
+    keptBarcodes = barcodes.loc[sorted([x[0] if len(x) == 1 else choice(x) for x
```
I would prefer to either take the barcode that is brighter or to create a new barcode that is placed at the mean of the x,y, and z positions of the duplicates so that the decoding can be as deterministic as possible.

emanuega commented on merlin/analysis/filterbarcodes.py:
```diff
@@ -316,7 +338,7 @@ def extreme_values(inputData: pandas.Series):
 
 
-class AdaptiveFilterBarcodes(decode.BarcodeSavingParallelAnalysisTask):
+class AdaptiveFilterBarcodes(FilterBarcodes):
```
This creates extra 'area_threshold', 'intensity_threshold', and 'distance_threshold' parameters in the parameter set for AdaptiveFilterBarcode that aren't relevant. I would prefer creating another parent class called AbstractFilterBarcodes that defines the shared parameters and can be used as the base class for both FilterBarcodes and AdaptiveFilterBarcodes.  The  _remove_zplane_duplicates function can all be moved to the utility module since I don't think the cleanup_across_z function is especially useful without this function.

emanuega commented on merlin/util/duplicatebcremoval.py:
```diff
@@ -0,0 +1,53 @@
+
+
+def cleanup_across_z(barcodes: pd.DataFrame, zPlanes: int, maxDist: float) \
+        -> pd.DataFrame:
```
Can we assert that all the barcodes have the same barcode_id and rename the function to remove_zplane_duplicates_single_barcodeid to make sure this function isn't accidentally misused?

seichhorn commented:
I've addressed your comments. I also changed the .item() call that was causing a future warning from pandas, and the sumsignal task to default to false for apply high pass - there are some issues of erroneous background generated from this at fov borders when the signal is low that I dont fully understand.

seichhorn commented:
Can you also merge 1.5 into master once this is in?
